[nav2_theta_star_planner] Theta* paths intersects lethal obstacles.

## Bug report

The `nav2_theta_star_planner/ThetaStarPlanner` produces paths that intersect lethal obstacles.

### Required Info

- Operating System: Ubuntu 22.04
- ROS2 Distro: Humble
- NAV2 Version: ros-humble-navigation2 1.1.20-1jammy.20251228.132604

### Steps to reproduce issue

1. Launch RViz2 and publish your static tf tree.
2. Use the provided `nav2_params.yaml` file to launch your `planner_server`, `bt_navigator` and `lifecycle_manager`.
3. Publish a map on the `/map` topic and use RViz2 to set a goal 2D pose.

### Expected behavior

The planner finds the cheapest path between the starting pose and the goal pose while avoiding obstacles.

### Actual behavior

The path intersects lethal obstacles.

### Configuration

`nav2_params.yaml`:

```
global_costmap:
  global_costmap:
    ros__parameters:
      global_frame: map
      robot_base_frame: base_link
      robot_radius: 0.5
      plugins: ["static_layer", "inflation_layer"]
      trinary_costmap: false
      lethal_cost_threshold: 99

      static_layer:
        plugin: "nav2_costmap_2d::StaticLayer"
        map_topic: /map
        subscribe_to_updates: false
        enabled: true
      
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        enabled: True
        cost_scaling_factor: 20.0
        inflation_radius: 1.0

planner_server:
  ros__parameters:
    use_sim_time: false
    expected_planner_frequency: 5.0
    costmap_update_timeout: 1.0
    allow_partial_planning: false
    planner_plugins: ["GridBased"]
    
    GridBased:
      plugin: "nav2_theta_star_planner/ThetaStarPlanner"
      how_many_corners: 8
      w_euc_cost: 1.0
      w_traversal_cost: 2.0

bt_navigator:
  ros__parameters:
    use_sim_time: false
    global_frame: map
    robot_base_frame: base_link
    odom_topic: odom
```

### Test Cases

- **Test 1: Lower Res Map with No Inflation Layer**
  - Size: 25 x 25 meters
  - Resolution: 1 m/px
  - Inflation Layer: None
  - Start: `x = -5.0 m`, `y = +11.0 m`
  - Goal: `x = +11.0 m`, `y = -8.0 m`

<img width="3045" height="3049" alt="Image" src="https://github.com/user-attachments/assets/f2177c4e-d5f1-40ca-9980-c2e8b923e5cf" />

```
[bt_navigator-5] [INFO] [1768318176.880994941] [bt_navigator]: Begin navigating from current location (-5.00, 11.00) to (11.00, -8.00)
[bt_navigator-5] [INFO] [1768318176.911177855] [bt_navigator]: Goal succeeded
```

**Result**: planner succeeds but path intersect lethal obstacles.

---

- **Test 2: Higher Res Map with No Inflation Layer**
  - Size: 125 x 125 meters
  - Resolution: 0.2 m/px
  - Inflation Layer: None
  - Start: `x = -5.0 m`, `y = +11.0 m`
  - Goal: `x = +11.0 m`, `y = -8.0 m`

<img width="3045" height="3049" alt="Image" src="https://github.com/user-attachments/assets/462daf66-c8c3-4ac3-ba0a-3ec740e2a70f" />

```
[bt_navigator-5] [INFO] [1768318176.880994941] [bt_navigator]: Begin navigating from current location (-5.00, 11.00) to (11.00, -8.00)
[bt_navigator-5] [INFO] [1768318176.911177855] [bt_navigator]: Goal succeeded
```

**Result**: planner succeeds but path intersect lethal obstacles.

---

- **Test 3: Higher Res Map with Inflation Layer**
  - Size: 125 x 125 meters
  - Resolution: 0.2 m/px
  - Inflation Layer: Yes (see **Configuration** section)
  - Start: `x = -5.0 m`, `y = +11.0 m`
  - Goal: `x = +11.0 m`, `y = -8.0 m`

<img width="3045" height="3049" alt="Image" src="https://github.com/user-attachments/assets/bf864373-9815-4430-981a-b213150a83a6" />

```
[bt_navigator-5] [INFO] [1768318176.880994941] [bt_navigator]: Begin navigating from current location (-5.00, 11.00) to (11.00, -8.00)
[bt_navigator-5] [INFO] [1768318176.911177855] [bt_navigator]: Goal succeeded
```

**Result**: planner succeeds but path intersect lethal regions of inflation layer.

---

- **Test 4: Narrower Map with No Inflation Layer**
  - Size: 25 x 25 meters
  - Resolution: 1.0 m/px
  - Inflation Layer: None
  - Start: `x = -5.0 m`, `y = +11.0 m`
  - Goal: `x = +11.0 m`, `y = -5.0 m`

<img width="3045" height="3049" alt="Image" src="https://github.com/user-attachments/assets/9e58685c-f8d9-4902-b939-84e11feffd47" />

```
[bt_navigator-5] [INFO] [1768317954.959984951] [bt_navigator]: Begin navigating from current location (-5.00, 11.00) to (11.00, -5.00)
[planner_server-4] [ERROR] [1768317954.960576225] [planner_server]: Could not generate path between the given poses
[planner_server-4] [WARN] [1768317954.960621343] [planner_server]: Planning algorithm GridBased failed to generate a valid path to (10.99, -5.06)
[planner_server-4] [WARN] [1768317954.960635590] [planner_server]: [compute_path_to_pose] [ActionServer] Aborting handle.
[bt_navigator-5] [WARN] [1768317954.990128094] [bt_navigator]: [navigate_to_pose] [ActionServer] Aborting handle.
[bt_navigator-5] [ERROR] [1768317954.990193396] [bt_navigator]: Goal failed
```

**Result**: planner fails to find viable path.

### Additional information

- For all tests, the `bt_navigator` uses custom behavior trees to only include the `planner_server` in the pipeline. Robot control is ignored.

--- a/nav2_theta_star_planner/src/theta_star.cpp
+++ b/nav2_theta_star_planner/src/theta_star.cpp
@@ -176,28 +176,31 @@ void ThetaStar::backtrace(std::vector<coordsW> & raw_points, const tree_node * c
   }
 }
 
-bool ThetaStar::losCheck(const int & x0, const int & y0, const int & x1, const int & y1, double & sl_cost) const {
+bool ThetaStar::losCheck(
+  const int & x0, const int & y0, const int & x1, const int & y1,
+  double & sl_cost) const
+{
   sl_cost = 0;
-  
+
   int dx = abs(x1 - x0), sx = (x0 < x1) ? 1 : -1;
   int dy = abs(y1 - y0), sy = (y0 < y1) ? 1 : -1;
-  int cx = x0, cy = y0, f = (dx >= dy) ? - dx / 2 : - dy / 2;
+  int cx = x0, cy = y0, f = (dx >= dy) ? -dx / 2 : -dy / 2;
 
   if (!isSafe(cx, cy)) {
     return false;
   }
-  
+
   if (dx >= dy) {
-    for (int i = 0; i <= dx; i++) {
+    for (int i = 0; i < dx; i++) {
       f += dy;
-      if (f > 0) {        
+      if (f > 0) {
         if (!isSafe(cx + sx, cy)) {
           return false;
         }
         if (!isSafe(cx, cy + sy)) {
           return false;
         }
-        
+
         cx += sx;
         cy += sy;
         f -= dx;
@@ -210,7 +213,7 @@ bool ThetaStar::losCheck(const int & x0, const int & y0, const int & x1, const i
       }
     }
   } else {
-    for (int i = 0; i <= dy; i++) {
+    for (int i = 0; i < dy; i++) {
       f += dx;
       if (f > 0) {
         if (!isSafe(cx + sx, cy)) {
@@ -219,7 +222,7 @@ bool ThetaStar::losCheck(const int & x0, const int & y0, const int & x1, const i
         if (!isSafe(cx, cy + sy)) {
           return false;
         }
-        
+
         cx += sx;
         cy += sy;
         f -= dy;
@@ -235,7 +238,6 @@ bool ThetaStar::losCheck(const int & x0, const int & y0, const int & x1, const i
 
   return true;
 }
-
 
 void ThetaStar::resetContainers()
 {

--- a/nav2_theta_star_planner/src/theta_star.cpp
+++ b/nav2_theta_star_planner/src/theta_star.cpp
@@ -176,62 +176,66 @@ void ThetaStar::backtrace(std::vector<coordsW> & raw_points, const tree_node * c
   }
 }
 
-bool ThetaStar::losCheck(
-  const int & x0, const int & y0, const int & x1, const int & y1,
-  double & sl_cost) const
-{
+bool ThetaStar::losCheck(const int & x0, const int & y0, const int & x1, const int & y1, double & sl_cost) const {
   sl_cost = 0;
+  
+  int dx = abs(x1 - x0), sx = (x0 < x1) ? 1 : -1;
+  int dy = abs(y1 - y0), sy = (y0 < y1) ? 1 : -1;
+  int cx = x0, cy = y0, f = (dx >= dy) ? - dx / 2 : - dy / 2;
 
-  int cx, cy;
-  int dy = abs(y1 - y0), dx = abs(x1 - x0), f = 0;
-  int sx, sy;
-  sx = x1 > x0 ? 1 : -1;
-  sy = y1 > y0 ? 1 : -1;
-
-  int u_x = (sx - 1) / 2;
-  int u_y = (sy - 1) / 2;
-  cx = x0;
-  cy = y0;
-
+  if (!isSafe(cx, cy)) {
+    return false;
+  }
+  
   if (dx >= dy) {
-    while (cx != x1) {
+    for (int i = 0; i <= dx; i++) {
       f += dy;
-      if (f >= dx) {
-        if (!isSafe(cx + u_x, cy + u_y, sl_cost)) {
+      if (f > 0) {        
+        if (!isSafe(cx + sx, cy)) {
           return false;
         }
+        if (!isSafe(cx, cy + sy)) {
+          return false;
+        }
+        
+        cx += sx;
         cy += sy;
         f -= dx;
+      } else {
+        cx += sx;
       }
-      if (f != 0 && !isSafe(cx + u_x, cy + u_y, sl_cost)) {
+
+      if (!isSafe(cx, cy, sl_cost)) {
         return false;
       }
-      if (dy == 0 && !isSafe(cx + u_x, cy, sl_cost) && !isSafe(cx + u_x, cy - 1, sl_cost)) {
-        return false;
-      }
-      cx += sx;
     }
   } else {
-    while (cy != y1) {
-      f = f + dx;
-      if (f >= dy) {
-        if (!isSafe(cx + u_x, cy + u_y, sl_cost)) {
+    for (int i = 0; i <= dy; i++) {
+      f += dx;
+      if (f > 0) {
+        if (!isSafe(cx + sx, cy)) {
           return false;
         }
+        if (!isSafe(cx, cy + sy)) {
+          return false;
+        }
+        
         cx += sx;
+        cy += sy;
         f -= dy;
+      } else {
+        cy += sy;
       }
-      if (f != 0 && !isSafe(cx + u_x, cy + u_y, sl_cost)) {
+
+      if (!isSafe(cx, cy, sl_cost)) {
         return false;
       }
-      if (dx == 0 && !isSafe(cx, cy + u_y, sl_cost) && !isSafe(cx - 1, cy + u_y, sl_cost)) {
-        return false;
-      }
-      cy += sy;
     }
   }
+
   return true;
 }
+
 
 void ThetaStar::resetContainers()
 {

--- a/nav2_theta_star_planner/src/theta_star.cpp
+++ b/nav2_theta_star_planner/src/theta_star.cpp
@@ -184,56 +184,31 @@ bool ThetaStar::losCheck(
 
   int dx = abs(x1 - x0), sx = (x0 < x1) ? 1 : -1;
   int dy = abs(y1 - y0), sy = (y0 < y1) ? 1 : -1;
-  int cx = x0, cy = y0, f = (dx >= dy) ? -dx / 2 : -dy / 2;
+  int cx = x0, cy = y0, e = dx - dy;
 
-  if (!isSafe(cx, cy)) {
-    return false;
+  while (cx != x1 || cy != y1) {
+    if (!isSafe(cx, cy, sl_cost)) {
+      return false;
+    }
+    int e2 = 2 * e;
+    if (e2 > -dy && e2 < dx) {
+      if (!isSafe(cx + sx, cy) || !isSafe(cx, cy + sy)) {
+        return false;
+      }
+      cx += sx;
+      cy += sy;
+      e += dx - dy;
+    } else if (e2 > -dy) {
+      cx += sx;
+      e -= dy;
+    } else {
+      cy += sy;
+      e += dx;
+    }
   }
 
-  if (dx >= dy) {
-    for (int i = 0; i < dx; i++) {
-      f += dy;
-      if (f > 0) {
-        if (!isSafe(cx + sx, cy)) {
-          return false;
-        }
-        if (!isSafe(cx, cy + sy)) {
-          return false;
-        }
-
-        cx += sx;
-        cy += sy;
-        f -= dx;
-      } else {
-        cx += sx;
-      }
-
-      if (!isSafe(cx, cy, sl_cost)) {
-        return false;
-      }
-    }
-  } else {
-    for (int i = 0; i < dy; i++) {
-      f += dx;
-      if (f > 0) {
-        if (!isSafe(cx + sx, cy)) {
-          return false;
-        }
-        if (!isSafe(cx, cy + sy)) {
-          return false;
-        }
-
-        cx += sx;
-        cy += sy;
-        f -= dy;
-      } else {
-        cy += sy;
-      }
-
-      if (!isSafe(cx, cy, sl_cost)) {
-        return false;
-      }
-    }
+  if (!isSafe(cx, cy, sl_cost)) {
+    return false;
   }
 
   return true;

--- a/nav2_theta_star_planner/src/theta_star.cpp
+++ b/nav2_theta_star_planner/src/theta_star.cpp
@@ -191,7 +191,7 @@ bool ThetaStar::losCheck(
       return false;
     }
     int e2 = 2 * e;
-    if (e2 > -dy && e2 < dx) {
+    if (e2 > -dy && e2 <= dx) {
       if (!isSafe(cx + sx, cy) || !isSafe(cx, cy + sy)) {
         return false;
       }
@@ -205,10 +205,6 @@ bool ThetaStar::losCheck(
       cy += sy;
       e += dx;
     }
-  }
-
-  if (!isSafe(cx, cy, sl_cost)) {
-    return false;
   }
 
   return true;

